src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+from typing import List, Optional, Tuple
+
+from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.snowflake_plan_node import Limit, LogicalPlan
+
+
+def infer_metadata(
+    source_plan: LogicalPlan,
+) -> Tuple[Optional[List[Attribute]], Optional[List[str]]]:
+    """
+    Infer metadata from the source plan.
+    Returns the metadata including attributes (schema) and quoted identifiers (column names).
+    """
+    from snowflake.snowpark._internal.analyzer.select_statement import (
+        Selectable,
+        SelectStatement,
+    )
+    from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+    from snowflake.snowpark._internal.analyzer.unary_plan_node import (
+        Filter,
+        Sample,
+        Sort,
+    )
+
+    attributes = None
+    quoted_identifiers = None
+    # If source_plan is a LogicalPlan, SQL simplfiier is not enabled
+    # so we can try to infer the metadata from its child (SnowflakePlan)
+    # When source_plan is Filter, Sort, Limit, Sample, metadata won't be changed
+    # so we can use the metadata from its child directly
+    if isinstance(source_plan, (Filter, Sort, Limit, Sample)):
+        if isinstance(source_plan.child, SnowflakePlan):
+            attributes = source_plan.child._attributes
+            quoted_identifiers = source_plan.child._quoted_identifiers
+    # If source_plan is a SelectStatement, SQL simplifier is enabled
+    elif isinstance(source_plan, SelectStatement):
+        # When source_plan._snowflake_plan is not None, `get_snowflake_plan` is called
+        # to create a new SnowflakePlan and `infer_metadata` is already called on the new plan.
+        if source_plan._snowflake_plan is not None:
```

Comment:
yes, we will cache the attribute at SelectStatement here, but we will do it in next PR (or if you think it's ok to include it this PR?)

Reply:
no, i don't think we want to do it in that pr, but could you remind me the example about why we need to cache it on the selectStatement (i recall the cache would lost under some case, but couldn't recall it exactly)

Reply:
Yes, it's needed for any dataframe after select, e.g., `df = session.sql("SELECT 1 AS a, 2 AS b").select("b")`

This is because after select, df._plan is resolved from df._select_statement (also, it's df._select_statement._snowflake_plan). When we call df.schema, attributes is cached at df._plan. Then if we call df.filter(...) when sql simplifier is enabled, df._select_statement._snowflake_plan is not copied to new SelectStatement, https://github.com/snowflakedb/snowpark-python/blob/2d61a4b65254d28279c9179d988f50ed75120f38/src/snowflake/snowpark/_internal/analyzer/select_statement.py#L1172-L1183, and df._plan is also not used anymore in new df._select_statement because of https://github.com/snowflakedb/snowpark-python/blob/2d61a4b65254d28279c9179d988f50ed75120f38/src/snowflake/snowpark/dataframe.py#L1336-L1340. It means we will lose attributes cached on df._plan and we need to cache on SelectStatement

Reply:
I went back to double check our conversion before[ ](https://snowflake.slack.com/archives/C03MJ5AA8CS/p1724431833197719) i think now i recall what is the problem now. This could still handle the selectStatement when flatten is not triggered, the cache on select statement is used to handle the case when flatten is triggered

Reply:
To be precise,
```
        if can_be_flattened:
            new = copy(self)
            new.from_ = self.from_.to_subqueryable()
            new.pre_actions = new.from_.pre_actions
            new.post_actions = new.from_.post_actions
            new.column_states = self.column_states
            new.where = And(self.where, col) if self.where is not None else col
            new._merge_projection_complexity_with_subquery = False
        else:
            new = SelectStatement(
                from_=self.to_subqueryable(), where=col, analyzer=self.analyzer
            )
```
when flatten is triggered, we can still handle it, because `from_` is carried over to next SelectStatement; when flatten is not triggered, we have to cache it on SelectStatement